### PR TITLE
Check bonsai connection; stop stage before moving; save camera parameters each trial

### PR DESCRIPTION
**Pull request by XX-Yin**
### Pull Request instructions:
- Please follow the [update protocol](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki/Software-Update-Procedures)
- Answer the questions below in detail. Your responses will be emailed to experimenters. 
- If the experimenters must do anything new, provide detailed step by step instructions on the [wiki](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki)
- Use markdown syntax in order for your comments to be rendered reliably in the email: "1." instead of "1)", use four spaces for indents.
- If you use the keyword "skip email" in the title, it will skip the email updates
- Merges from "production_testing" into "main" should use the keyword "production merge" in the title for reliable indexing of updates
  
### Describe changes:
1. Check bonsai connections when using the camera and laser calibration. 
2. Stop the stage before moving. 
3. Save camera parameters each trial. 
4. Enable the optogenetics2 node. 
5. Add ReconnectBonsai button. 

### What issues or discussions does this update address?
1.  The GUI will crash If we run some functions without Bonsai connected to the GUI. It's a good to check the connection before running these functions.
2. If the stage has not stopped, the stage will not move when we click the move button.
3. It's better to save the camera parameters each trial. Previously, it only saved the parameter of the last trial. 

### Describe the expected change in behavior from the perspective of the experimenter
No
### Describe the outcome of testing this update on a rig in 447
I tested them in the ephys rig 3. It works.

### CODE REVIEW

**Comment by alexpiet** [comment]
What does the ReconnectBonsai button do?

File: src/foraging_gui/Foraging.py [hunk 1]
```diff
@@ -368,6 +370,16 @@ def _ConnectBonsai(self):
                 self.WarningLabelInitializeBonsai.setStyleSheet("color: red;")
                 self.InitializeBonsaiSuccessfully=0
 
+    def _ReconnectBonsai(self):
+        '''Reconnect bonsai'''
+        if self.InitializeBonsaiSuccessfully==1:
+            self.client.close()
```
Comment (alexpiet) [inline src/foraging_gui/Foraging.py:376]: Why would we want to close the clients and reconnect?

**Comment by XX-Yin** [comment]
> What does the ReconnectBonsai button do?

It will disconnect the Bonsai from the GUI and connect them again.

File: src/foraging_gui/Foraging.py (hunk 1 above)
Comment (XX-Yin) [reply src/foraging_gui/Foraging.py:376]: In some cases, the connection between GUI and Bonsai is interrupted by mistake and we can reconnect them. I also use it for testing when I want to change the bonsai code while the connection is already established so I don't need to reopen the GUI.

File: src/foraging_gui/Dialogs.py [hunk 2]
```diff
@@ -458,6 +458,9 @@ def _LoadCaliPar(self):
 
     def _StartCalibratingLeft(self):
         '''start the calibration loop of left valve'''
+        self.MainWindow._CheckBonsaiConnection()
+        if self.MainWindow.InitializeBonsaiSuccessfully==0:
+            return
```
Comment (alexpiet) [inline src/foraging_gui/Dialogs.py:463]: What happens if we return? Does nothing happen? Should the user be alerted in some way?

Comment (XX-Yin) [reply src/foraging_gui/Dialogs.py:463]: A warning "Please open Bonsai!" will appear in the main window

File: src/foraging_gui/Foraging.py [hunk 3]
```diff
@@ -2080,12 +2094,16 @@ def _set_metadata_enabled(self, enable: bool):
         self.Experimenter.setEnabled(enable)
         self.Tower.setEnabled(enable)
 
+    def _CheckBonsaiConnection(self):
```
Comment (alexpiet) [inline src/foraging_gui/Foraging.py:2097]: This doesn't check the connection, it only establishes a connection if `InitializeBonsaiSuccessfully` = 0. Shouldn't it always check?

Comment (XX-Yin) [reply src/foraging_gui/Foraging.py:2097]: Yes. How about we merge this first? And I will create a separate pull request to solve this issue. This problem only occurs when the Bonsai and GUI are connected and then disconnected.

**Review by alexpiet — APPROVED** [review]

File: src/foraging_gui/Foraging.py (hunk 3 above)
Comment (alexpiet) [reply src/foraging_gui/Foraging.py:2097]: ok

Comment (alexpiet) [reply src/foraging_gui/Foraging.py:2097]: https://github.com/AllenNeuralDynamics/dynamic-foraging-task/issues/111

**Pull request merged**